### ISSUE TITLE
Grille pain

### ISSUE THREAD

**Issue by michelollivier** [patrickschur/language-detection#1]
Guten Tag,

I continue in english ;)

I compared all current language detectors based on  sequence of N-grams and **your solution is the best implementation**.

However I have to translate very short sequences of words. For example the words 'Grille pain' which means toaster in in english returns

```
array(6) {
  ["it"]=>
  float(0.54711111111111)
  ["fr"]=>
  float(0.54633333333333)
  ["en"]=>
  float(0.506)
  ["de"]=>
  float(0.49488888888889)
  ["nl"]=>
  float(0.49)
  ["es"]=>
  float(0.43466666666667)
}
```
It's almost good! it wins on fr by a difference 0.000777. 

So, I added words 'Grille pain' in fr file, set trainer and new result gives:

```
array(6) {
  ["fr"]=>
  float(0.54944444444444)
  ["it"]=>
  float(0.54711111111111)
  ["en"]=>
  float(0.506)
  ["de"]=>
  float(0.49488888888889)
  ["nl"]=>
  float(0.49)
  ["es"]=>
  float(0.43466666666667)
}
```
It's good. fr wins on it by a difference of 0.002333.

So my questions , 

1- Can I populate language files so that I can be sure that the first occurence wins with a very significant difference.

2- If yes to previous question, is the detection process slower if language files are very big?

3- I can see that, in fr file, you have put french declaration of rights. I dont think that a 200 years old text represents very well current french language. Is there somewhere some data which may be more accurate?

Thanks for your great job.

Best Regards
Michel

**Comment by patrickschur** [comment]
First off thank you so much, for the star.  It makes me happy to hear that, that my language detector is the best one you ever used. :smile:

Language detectors based on N-grams are actually not made for small sentences like yours.
It works surprisingly well, however, they are really powerful only at longer sentences/words.
If you want to achieve better results, you have to change the language files, as you have already done.
To improve the performance of the script you can also remove languages or use the constructor.

To come to your questions:
1) How do I understand this? Did you mean something like this?
```text
array(6) {
  ["fr"]=>
  float(1)
  ["it"]=>
  float(0.54711111111111)
  ["en"]=>
  float(0.506)
  ["de"]=>
  float(0.49488888888889)
}
```
The problem is that there are some languages that are not very different from each other. If you use small sentences. So no,it is not possible at the current time. Just try a bit to play with the trainer class and the language files. I will also look at it again later.

2) No it is not. :) The trainer class will only use the best 300 N-grams of the language. If you don't change this number or add more language files it will not affect the performance. Only creating the N-grams is slower. However, the creation of N-grams must be done only once. 
Summary: The training phase will be slower but the detection phase remains the same.

3) I'm sorry I don't speak French :cry:. It's hard to find files for more than 100 languages. This is the reason why all files are the same. In addition the results are very good. If you want to get better results, try to use wikipedia articles.
Maybe I'll take this into consideration.

Are you German? :)

I thank you too.
Patrick Schur

**Comment by jpmurray** [comment]
Hey @patrickschur , I see you've updated the french text (or and others!) source with the Universal Declaration of Human Rights and I wanted to say: good call ;-)

**Comment by patrickschur** [comment]
I didn't changed the french text. I fixed a bug in the algotihm, increased the amount of n-grams which are compared and improved the speed. Which produced better results.
But thanks. :+1:

**Comment by jpmurray** [comment]
AH! I though it was new because of one of the original comment saying it was the "french declaration of rights" of over 200 years old, and you changed it for the official text from the UN, which is not that old and has over 500 translations! :)

My bad @patrickschur ! :D

**Comment by patrickschur** [comment]
@jpmurray No problem! All texts I'm using are based on the [Universal Declaration of Human Rights](http://www.unicode.org/udhr/). When you are from Canada, you speak french, right? Could you check out and tell me if the translation from the [UN](http://www.ohchr.org/EN/UDHR/Pages/Language.aspx?LangID=frn) is better than the other one? :)

**Comment by jpmurray** [comment]
@patrickschur just made a quick overview, but it seems pretty much the same to me!

**Comment by patrickschur** [comment]
@jpmurray Thanks for checking. I will appreciate it. ;-)
